kubernetes/pods.go

```diff
@@ -0,0 +1,85 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apisv1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+)
+
+const resyncInterval = 1 * time.Minute
+
+// PodInformer is a event handler for Pod events registered to, that builds a local list of valid and relevant pods
+// and sends an event to the endpoint channel, triggering a resync of the targets.
+func (a *Adapter) PodInformer(ctx context.Context, endpointChan chan<- []string) error {
+	podEndpoints := sync.Map{}
+
+	log.Infof("Watching for Pods with labelselector %s in namespace %s", a.cniPodLabelSelector, a.cniPodNamespace)
+	factory := informers.NewSharedInformerFactoryWithOptions(a.clientset, resyncInterval, informers.WithNamespace(a.cniPodNamespace),
+		informers.WithTweakListOptions(func(options *apisv1.ListOptions) { options.LabelSelector = a.cniPodLabelSelector }))
+
+	informer := factory.Core().V1().Pods().Informer()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(_, newResource interface{}) {
+			pod, ok := newResource.(*corev1.Pod)
+			if !ok {
+				log.Printf("cannot cast object to corev1.Pod %v", newResource)
+				return
+			}
+			switch {
+
+			case pod.DeletionTimestamp != nil:
+				name, exists := podEndpoints.LoadAndDelete(pod.Name)
+				if !exists {
+					return
+				}
+				log.Infof("Deleted pod: %s IP: %s", pod.Name, name)
+				queueEndpoints(&podEndpoints, endpointChan)
+
+			case pod.Status.ContainerStatuses != nil && len(pod.Status.ContainerStatuses) > 0 &&
+				pod.Status.ContainerStatuses[0].State.Running != nil && pod.Status.PodIP != "":
+				if _, isStored := podEndpoints.LoadOrStore(pod.Name, pod.Status.PodIP); isStored {
+					return
+				}
+				log.Infof("New discovered pod: %s IP: %s", pod.Name, pod.Status.PodIP)
+				queueEndpoints(&podEndpoints, endpointChan)
+			}
+		},
+	})
+
+	factory.Start(ctx.Done())
+	if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
+		return fmt.Errorf("Timed out waiting for caches to sync")
+	}
```

Comment:
Please do not resolve until we come to a conclusion.

> listers/informers must be accessed and set up prior to starting the shared informers

In the ticket you are referencing the problem is different, they start informer and only then call `AddHandler`. 
I propose a way to postpone sending updates until cache is updated if that is needed. 

> factory.Start is non-blocking, required for caches to built up. Then calling informer.Run which is blocking. Thats the canonical way

we can wait for `ctx.Done()` after calling  `factory.Start`

```go
var sendUpdates int64 = 0
...
informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
		UpdateFunc: func(_, newResource interface{}) {
                        ...
                        // postpone sending endpoint updates until cache is ready
		        if atomic.LoadInt64(&sendUpdates) != 0 {
		    		queueEndpoints(&podEndpoints, endpointChan)
			}
			....
		},
		...
)
...
factory.Start(ctx.Done())
factory.WaitForCacheSync(ctx.Done())
atomic.StoreInt64(&sendUpdates, 1)
<-ctx.Done()
return ctx.Err()
```

Reply:
your logic is flawed, it skips the first event!

Reply:
You can see in the unit test that the informer.Run is though required

Reply:
>  it skips the first event! 
```go
...
atomic.StoreInt64(&sendUpdates, 1)
// trigger an update
queueEndpoints(&podEndpoints, endpointChan)
```

Reply:
doesnt work see above, lets use only one thread to discuss

Reply:
> doesnt work see above

Could you elaborate what does not work or link to the comment?

The idea is that in the `AddEventHandler` we accumulate updates but only send them when `sendUpdates` is set.
We set sendUpdates only when waiting for the cache is complete and manually trigger `queueEndpoints(&podEndpoints, endpointChan)` after setting the flag:

```go
func (a *Adapter) PodInformer(ctx context.Context, endpointChan chan<- []string) error {
	podEndpoints := sync.Map{}
	var sendUpdates int64 = 0

	log.Infof("Watching for Pods with labelselector %s in namespace %s", a.cniPodLabelSelector, a.cniPodNamespace)
	factory := informers.NewSharedInformerFactoryWithOptions(a.clientset, resyncInterval, informers.WithNamespace(a.cniPodNamespace),
		informers.WithTweakListOptions(func(options *apisv1.ListOptions) { options.LabelSelector = a.cniPodLabelSelector }))

	informer := factory.Core().V1().Pods().Informer()
	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
		UpdateFunc: func(_, newResource interface{}) {
			pod, ok := newResource.(*corev1.Pod)
			if !ok {
				log.Printf("cannot cast object to corev1.Pod %v", newResource)
				return
			}
			switch {

			case pod.DeletionTimestamp != nil:
				name, exists := podEndpoints.LoadAndDelete(pod.Name)
				if !exists {
					return
				}
				log.Infof("Deleted pod: %s IP: %s", pod.Name, name)
				if atomic.LoadInt64(&sendUpdates) != 0 {
		    			queueEndpoints(&podEndpoints, endpointChan)
				}

			case pod.Status.ContainerStatuses != nil && len(pod.Status.ContainerStatuses) > 0 &&
				pod.Status.ContainerStatuses[0].State.Running != nil && pod.Status.PodIP != "":
				if _, isStored := podEndpoints.LoadOrStore(pod.Name, pod.Status.PodIP); isStored {
					return
				}
				log.Infof("New discovered pod: %s IP: %s", pod.Name, pod.Status.PodIP)
				if atomic.LoadInt64(&sendUpdates) != 0 {
		    			queueEndpoints(&podEndpoints, endpointChan)
				}
			}
		},
	})

	factory.Start(ctx.Done())
	if !cache.WaitForCacheSync(ctx.Done(), informer.HasSynced) {
		return fmt.Errorf("Timed out waiting for caches to sync")
	}

	// 
	// I guess this comment should rather explain that listing triggers AddEventHandler 
	// 
	// initial verification if ingress pods for given specs can be found, otherwise the informer would fail silently
	for {
		list, err := factory.Core().V1().Pods().Lister().List(labels.Everything())
		if err == nil && len(list) > 0 {
			break
		}
		log.Errorf("Error listing Pods with labelselector %s in namespace %s: %v", a.cniPodNamespace, a.cniPodLabelSelector, err)
		time.Sleep(resyncInterval)
	}

	atomic.StoreInt64(&sendUpdates, 1)
	// force an initial update
	queueEndpoints(&podEndpoints, endpointChan)

	informer.Run(ctx.Done())
	return nil
}
```

Reply:
please see https://github.com/zalando-incubator/kube-ingress-aws-controller/pull/461#discussion_r789661577